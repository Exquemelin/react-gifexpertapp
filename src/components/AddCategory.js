import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories })  => {

    // Cremoas una constante que iremos cambiando según varíen los State
    const [inputValue, setInputValue] = useState(''); // Tenemos que poner '' para que no sea un undefined y nos de un warning

    // Usamos un método para modificar la variable inputValue
    const handleInputChange = ( e ) => {
        // console.log(e.target.value);
        setInputValue(e.target.value);

        // Este Console.log es para las pruebas
        console.log('handleInputChange fue lanzado');
    }

    const handleSubmit = ( e ) => {

        // Prevenimos que al hacer submit recarge el formulario
        e.preventDefault();

        // Console.log para las pruebas
        console.log('handleSubmit lanzado', inputValue);

        // Validamos lo que nos entra por el input
        if ( inputValue.trim().length > 2 ) {

            // Activamos el setCategories que nos entra por las props
            setCategories( cats => [inputValue, ...cats] );
            // Y vaciamos el inputValue
            setInputValue('');

        }


    }
    
    // Usamos un form en lugar de un fragment, ya que también agrupa elementos y queremos hacer un formulario
    return (
        <form onSubmit={ handleSubmit }>
            <p> { inputValue } </p> 
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )

}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}