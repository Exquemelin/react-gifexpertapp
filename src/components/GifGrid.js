// useEffect nos permite ejecutar cógido de forma condicional
import React from 'react' 
import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';


import { GifGridItem } from './GifGridItem';



export const GifGrid = ({ category }) => {

    // const [images, setImages] = useState([]);

    // Importamos el hook y lo desestructuramos
    const { data:images, loading } = useFetchGifs( category );


    // // Condicionamos el disparo del método getGifs
    // useEffect( () => {
    //     getGifs( category )
    //     // .then( imgs => setImages( imgs ) );
    //     .then( setImages ); // al tener un solo argumento, se puede hacer más corta
    // }, [ category ]); // Si añadimos [] solo se ejecuta cuando es renderizado por primera vez
    // // Si añadimos [category] se vuelve a renderizar


    return (
        <>
            <h3 className="animate__animated animate__fadeIn"> { category } </h3>

            { loading && <p className="animate__animated animate__flash">Loading.....</p>}

            <div className="card-grid">

                {
                    images.map( gif => (
                        // <li key={ gif.id}> {gif.title} </li>
                        // console.log({ ...gif })
                        <GifGridItem 
                            key={ gif.id }
                            { ...gif }
                        />
                    ))
                }

            </div>
        </>
    )
}
