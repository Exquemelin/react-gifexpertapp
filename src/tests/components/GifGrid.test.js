import { shallow } from "enzyme";
import React from 'react';
import '@testing-library/jest-dom';


import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
// Esto simula la llamada al elemento cuya ruta le marcamos. En este caso el useFetchGifs
jest.mock('../../hooks/useFetchGifs');




describe('Pruebas con el componente <GifGrid />', () => {

    // Definimos una variable con una categoría
    const category = 'Gallactica';

    // Declaramos una variable para almacenar el código
    // let wrapper = shallow(<GifGrid category={ category }/>);

    test('debe validar el snapshot', () => {

        // Le decimos que useFetchGifs va a retornar en la simulación la data vacía, y el loading true, que son los datos iniciales
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });

        const wrapper = shallow(<GifGrid category={ category } />);
        // Comprobamos que coindicina con el snapshot
        expect( wrapper ).toMatchSnapshot();
        
    });

    test('debe de mostrar items cuando se cargan imágenes con useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'http//localhost/cualquier/cosa.jpg',
            title: 'Cualquier Cosa'
        },{
            id: '123',
            url: 'http//localhost/cualquier/cosa.jpg',
            title: 'Cualquier Cosa'
        }]
        
        // Le pasamos datos para la simulación
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });
        
        const wrapper = shallow(<GifGrid category={ category }/>);

        // Comprobamos que no haya ningún párrafo 'p'
        expect( wrapper.find('p').exists() ).toBe( false );

        // Comprobamos cuantos elementos se renderizaron
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );
        
    });
    
    
    
});
