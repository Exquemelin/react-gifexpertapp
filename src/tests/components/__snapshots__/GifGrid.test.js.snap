// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con el componente <GifGrid /> debe validar el snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    Gallactica
     
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading.....
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
